include/verilated_trace_imp.cpp

```diff
@@ -90,6 +90,221 @@ class VerilatedTraceCallInfo {
+    // required, but don't allocate more than 8 buffers.
+    if (m_numTraceBuffers < 8) {
+        // Allocate a new buffer if none is available
+        if (!m_buffersFromWorker.tryGet(bufferp)) {
+            m_numTraceBuffers++;
+
+            // We need to be able to store a new value for each signal, which is
+            // 'nextCode()' entries, plus up to 3 more words of metadata per signal,
+            // Plus fixed overhead of 1 for a termination flag and 2 for a time
+            // stamp update.
+            bufferp = new VerilatedTraceEntry[nextCode() + numSignals() * 3 + 3];
+        }
+    } else {
+        // Block until a buffer becomes available
+        bufferp = m_buffersFromWorker.get();
+    }
+    return bufferp;
+}
+
+template <> void VerilatedTrace<VL_DERIVED_T>::waitForBuffer(const VerilatedTraceEntry* buffp) {
+    // Collect buffers from worker and stash them until we get the one we want
+    std::deque<VerilatedTraceEntry*> stash;
+    do { stash.push_back(m_buffersFromWorker.get()); } while (stash.back() != buffp);
+    // Now put them back in the queue, in the original order
+    while (!stash.empty()) {
+        m_buffersFromWorker.put_front(stash.back());
+        stash.pop_back();
+    }
+}
+
+//=========================================================================
+// Worker thread
+
+#if 0
+#define VL_TRACE_THREAD_DEBUG(msg) std::cout << "TRACE THREAD: " << msg << std::endl
+#else
+#define VL_TRACE_THREAD_DEBUG(msg)
+#endif
+
+template <> void VerilatedTrace<VL_DERIVED_T>::workerThreadMain() {
+    while (true) {
+        VerilatedTraceEntry* const bufferp = m_buffersToWorker.get();
+
+        VL_TRACE_THREAD_DEBUG("");
+        VL_TRACE_THREAD_DEBUG("Got buffer");
+
+        const VerilatedTraceEntry* readp = bufferp;
+
+        vluint32_t cmd;
+        unsigned bits;
+        vluint32_t* oldp;
+        vluint64_t newBits;
+
+        while (true) {
+            cmd = (readp++)->cmd;
+
+            switch (cmd & ~0xFFU) {
+                //==================================================================
+                // CHG_* commands
+            case VerilatedTraceCommand::CHG_BIT:
+                VL_TRACE_THREAD_DEBUG("Command CHG_BIT");
+                chgBitImpl((readp++)->oldp, cmd & 1);
+                continue;
+            case VerilatedTraceCommand::CHG_BUS:
+                VL_TRACE_THREAD_DEBUG("Command CHG_BUS");
+
+                oldp = (readp++)->oldp;
+                newBits = (readp++)->newBits;
+
+                // Bits stored in bottom byte of command
+                switch (cmd & 0xFFU) {
+                case 2: chgBusImpl<2>(oldp, newBits); continue;
+                case 3: chgBusImpl<3>(oldp, newBits); continue;
+                case 4: chgBusImpl<4>(oldp, newBits); continue;
+                case 5: chgBusImpl<5>(oldp, newBits); continue;
+                case 6: chgBusImpl<6>(oldp, newBits); continue;
```

Comment:
I don't think this will be a performance win, you're better off passing cmd&0xFF to the non-templated single function jump table approach, because cmd isn't a constant the compiler can't know which to call in advance, then you have all of these largeish functions some might have fallen out of the cache.  In contrast the non-templated has just a single unpredictable branch, equal to what you already have.

Reply:
Yes, that is quite clear. It is this way because the Verilator generated code at the moment is the same whether the `--trace-threads` is used or not (i.e.: `--trace-threads` right now is implemented in libverilated only + a few defines in the generated makefile). This means the function signatures must match and I did not want to slow down the single threaded case without doing more rigorous performanc analysis regardign the jump table approach for small signals which I still fear would be measurable. My intention was to do that in a subsequent patch, to keep patches smaller and focusing on a single issue where possible (single threaded vs multi threaded tracing in this case). I am happy to do this in this patch if you would like me to, I just thought you would appreciate easier to review patches :wink:. BTW please let me know if you prefer a different style.

For optimal performance long term, the function signatures of `full*` and `chg*` will have to change between single vs multi-threaded tracing, but for example the FST writer's own thread compilcates things a bit, as for optimal performance with `--trace-fst --trace-threads 1` still uses the "single threaded API" in the emitted `traceChg`, but `--trace-fst --trace-threads 2` or `--trace --trace-threads 1` would use the "multi threaded API", so it's a bit less than straigth forward to reason about. In the long term maybe we can eliminate the inefficiency of the FST API so the FST writer's own thread won't be necessary, but right now it is very helpful.

In general the performance of the code in the worker thread right now is secondary as simulation speed (i.e.: user perceived latency/wall clock) is bottlenecked by eval+traceChg which runs on the main thread. Not this will likely change as I imporove things, but is true right now.

Reply:
Seems right & a separate optimization patch is reasonable & appreciated.